### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by kimoyerr**
Added model for GPT 
And some checks for forward pass

## Release Notes

- **Configuration**
  - Updated configuration settings for model training, including new parameters for device, batch size, learning rate, and validation intervals.

- **Model Development**
  - Introduced new model components including Multi-Head Attention, FeedForward, TransformerBlock, and GPT classes.
  - Added utility functions for computing positional embeddings and learning rate scheduling.

- **Dependencies**
  - Added `torchinfo` package (version >=1.8.0) and `wandb` package (version >=0.19.3).

- **Training Updates**
  - Adjusted data loading process with reduced batch size.
  - Implemented dynamic sequence length configuration.
  - Introduced a functional forward pass for model predictions.
  - Added loss masking for padded tokens during training.
  - Enhanced logging capabilities using Weights & Biases.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -12,6 +12,7 @@ dependencies = [
     "pytest>=8.3.4",
     "seaborn>=0.13.2",
     "torchdata>=0.10.1",
+    "torchinfo>=1.8.0",
```
[the remark this answers is not in the document]
Comment (kimoyerr) [reply pyproject.toml:15]: Leave it on for now

File: train.py [hunk 2]
```diff
@@ -40,14 +45,23 @@
 model_name = f'facebook/esm1v_t33_650M_UR90S_{model_seed}'
 tokenizer = EsmTokenizer.from_pretrained(model_name)
 # tokenizer.add_tokens(['<bos>', '<eos>'])
-data_loader = build_hf_data_loader('ecoli_protein_train', dataloader_dir_path, "train", "Sequence", tokenizer, batch_size=24, seq_len=256, world_size=1, rank=0, infinite=True)
-val_data_loader = build_hf_data_loader('ecoli_protein_val', dataloader_dir_path, "validation", "Sequence", tokenizer, batch_size=24, seq_len=256, world_size=1, rank=0, infinite=True)
+data_loader, num_train_seqs = build_hf_data_loader('ecoli_protein_train', dataloader_dir_path, "train", "Sequence", tokenizer, batch_size=4, seq_len=model_cfg.max_seq_len, world_size=1, rank=0, infinite=True)
+val_data_loader, num_val_seqs = build_hf_data_loader('ecoli_protein_val', dataloader_dir_path, "validation", "Sequence", tokenizer, batch_size=4, seq_len=model_cfg.max_seq_len, world_size=1, rank=0, infinite=False)
```
[the remark this answers is not in the document]
Comment (kimoyerr) [reply train.py:49]: Outdated
